### ISSUE TITLE
Poor cypher performance when registering AuthSession

### ISSUE THREAD

**Issue by adelefelicia** [indykite/neo4j-graph-tool-core#77]
**Created by: @bayegan Jun 24 2021 at 17:34 CET**
Last updated: Apr 22 2022 at 06:49 CET
ClickUp ID: IK-1147

Looking at trace in dev, this cypher performs specially bad. 

https://github.com/indykite/jarvis-proto/blob/develop/pkg/cypher/session.go

See the latest perf test:
https://drive.google.com/drive/folders/1l2MVeryPuhHSfXa4TY-3Ywzda8P925Jm

https://console.cloud.google.com/traces/list?project=jarvis-dev-268314&start=1619337895650&end=1619338279671&tid=098f119586741210ee83427a2fece148 

Investigation
Validating resource consumption during tests shows that there is no issue neither with CPU nor with RAM availability.
There is no problem with GC, logs shows that DB works properly from Java perspective.
Memory settings has been adjusted to the values proposed by neo4j-admin tool:
dbms.memory.heap.initial_size=5100m
dbms.memory.heap.max_size=5100m
dbms.memory.pagecache.size=6900m

Looks like he problem with registration performance is related to the db locks. There are plenty of registrations happening at the same time and each of them requires exclusive lock on Tenant (to create relationship)
"Blocked by: [neo4j-transaction-107214]"
About 50-60% of all running transactions are blocked by a single exclusive lock -> all transactions have to be executed in sequence! Disks IOPS are not problem, switching to high IOPS SSD disk didn't change anything.

Locks problem temporarily has been addressed (for further testing only) by introducing Container nodes between DT and Tenant and between Application and Session. 

We are still observing long response times but it seems this time it might be coming from backend services. We have observed 100% CPU utilisation during tests (in dev currently running 5 PODs with hermes service). 
Before ew increase number of PODs we are introducing some additional traces to hopefully narrow down what is the bottleneck in the services.

 

In addition to improvements to the how data is stored in DB the start email workflow had a significant impact on performance (was only able to see it once issue with DB performance was gone)

 

Making that call async we were able to get the following results:

**Comment by adelefelicia** [comment]
**Commented by: @bayegan Jul 06 2021 at 22:37 CET**

Duplicate of hqaefp
